### DISCUSSION TITLE
cannot run podman API examples: server API version is too old (Ubuntu 22.04LTS)

### DISCUSSION THREAD

**Discussion by thediveo** [containers/podman#15543] [Q&A]
- I've followed https://podman.io/getting-started/ to install podman from Ubuntu's official 22.04LTS repositories.
- I've then followed https://podman.io/getting-started/installation to install the necessary development libraries/packages.
- I've then followed https://github.com/containers/podman/tree/main/pkg/bindings with the first minimal example (which doesn't compile) and got it running after some corrections.

The example then "crashes":

```
go run -exec sudo main.go
unable to connect to Podman socket: server API version is too old. Client "4.0.0" server "3.4.4"
exit status 1
```

How do I build a podman REST API client that can connect to both 3.x and 4.x APIs?

And why do I need to install a lot of highly special development library which aren't required by any REST API client but only when building Podman itself? Is there some build flag to avoid all the unnecessary stuff and cut right through to the client stuff needed?

**Comment by mheon** [comment]
I don't believe that is possible without vendoring two versions of the API bindings. There were several breaking changes from 3.0 to 4.0 in the network endpoints, that make it impossible to support both simultaneously with the same bindings.

**Reply by thediveo** [reply]
Thank you very much for your answer! After reading the source code for [`pkg/bindings/connection.go`](https://github.com/containers/podman/blob/main/pkg/bindings/connection.go#L182) I've found that the contract seems to be that an older client may talk to a newer server using its protocol version, as long as the server accepts the older version and handles it correctly.

I'm now trying with the v3 client; however, that causes other problems, such as the need to pin the dependency for `github.com/containernetworking/cni` to `0.8.1`. I also found that [`DoRequest`](https://github.com/containers/podman/blob/v3.4/pkg/bindings/connection.go#L309) in v3 doesn't even use any context, so it is impossible to cancel or deadline any client request.

This really is messy, more so when compared to The Whale. Never had such problems with its Go client.
